pkg/router/router.go

```diff
@@ -164,152 +267,80 @@ func registerPropertiesEndpoint(r endpoints.Resource, container *restful.Contain
 	container.Add(wsDefaults)
 }
 
-// Back-end extension: Requests to the URL are passed through to the Port of the Name service (extension)
-// "label: tekton-dashboard-extension=true" in the service defines the extension
-// "annotation: tekton-dashboard-endpoints=<URL>" specifies the path for the extension
+// Extension is the back-end representation of an extension. A service is an
+// extension when it is in the dashboard namespace with the dashboard label
+// key/value pair. Endpoints are specified with the extension URL annotation
 type Extension struct {
-	Name           string `json:"name"`
-	URL            string `json:"url"`
-	Port           string `json:"port"`
-	DisplayName    string `json:"displayname"`
-	BundleLocation string `json:"bundlelocation"`
+	Name           string   `json:"name"`
+	URL            *url.URL `json:"url"`
+	Port           string   `json:"port"`
+	DisplayName    string   `json:"displayname"`
+	BundleLocation string   `json:"bundlelocation"`
+	endpoints      []string
 }
 
-// Informer only receives services and must be able to map to WebService/Extension
-var extensionMutex *sync.RWMutex = new(sync.RWMutex)
-var extensionWebService *restful.WebService
-var uidExtensionMap map[string]Extension = make(map[string]Extension)
-
-// Registers the WebService responsible for proxying to all extensions and also to getAll extensions
-func registerExtensions(container *restful.Container) {
-	logging.Log.Info("Adding API for Extensions")
-	extensionWebService = new(restful.WebService)
-	extensionWebService.SetDynamicRoutes(true)
-	extensionWebService.
-		Path(ExtensionRoot).
-		Consumes(restful.MIME_JSON).
-		Produces(restful.MIME_JSON)
-	extensionWebService.Route(extensionWebService.GET("").To(getAllExtensions))
-	container.Add(extensionWebService)
+// newExtension returns a new extension
+func newExtension(extService *corev1.Service) *Extension {
+	port := getServicePort(extService)
+	url, _ := url.ParseRequestURI(fmt.Sprintf("http://%s:%s", extService.Spec.ClusterIP, port))
+	return &Extension{
+		Name:           extService.ObjectMeta.Name,
+		URL:            url,
+		Port:           port,
+		DisplayName:    extService.ObjectMeta.Annotations[ExtensionDisplayNameKey],
+		BundleLocation: extService.ObjectMeta.Annotations[ExtensionBundleLocationKey],
+		endpoints:      getExtensionEndpoints(extService.ObjectMeta.Annotations[ExtensionURLKey]),
+	}
 }
 
-func GetExtensions() []Extension {
-	extensions := []Extension{}
-	extensionMutex.RLock()
-	for _, e := range uidExtensionMap {
-		extensions = append(extensions, e)
-	}
-	extensionMutex.RUnlock()
-	return extensions
+// MarshalJSON marshals the Extension into JSON.
+func (e Extension) MarshalJSON() ([]byte, error) {
+	type Alias Extension
```

Comment:
What does this do and how come it's now needed?

Reply:
We used to construct the url each time we handled proxy requests. Instead, now when extension registration happens, I construct the URL. However, url.URL is not marshalled as expected (see: https://play.golang.org/p/6ffOLym8hz2). This overrides the default MarshalJSON method so that url.URL marshals as expected with the old payload.